internal/metamorphic/crossversion/crossversion_test: TestMetaCrossVersion failed

internal/metamorphic/crossversion/crossversion_test.TestMetaCrossVersion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/16404922?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/16404922?buildTab=artifacts#/) on refs/heads/master @ [791b3749feb5](https://github.com/cockroachdb/pebble/commits/791b3749feb5):


```
        iter23.Next("")
        iter27.SetBounds("bwpsxjykc@15", "bwpsxjykc@15")
        iter25.First()
        snap18.Get("uikp@15")
        snap18.Get("bklzg@46")
        iter25.Prev("")
        iter23.First()
        iter23.Prev("")
        iter24.SeekLT("wzro@8", "")
        iter25.Prev("")
        db1.IngestExternalFiles(external0, "pfhlmp" /* start */, "xfdzvesqzfqu" /* end */, "" /* syntheticSuffix */, "" /* syntheticPrefix */)
        snap18.Get("ipyfbzjjkccz@13")
        iter24.SeekGE("uikp@4", "ukqngnbb")
        db1.SingleDelete("wintgpfpc", false /* maybeReplaceDelete */)
        iter21.NextPrefix()
        snap18.Get("njigqiwnfw")
        db1.IngestExternalFiles(external0, "xfdzvesqzfqu" /* start */, "ybglsmtq" /* end */, "" /* syntheticSuffix */, "" /* syntheticPrefix */)
        iter27.Next("")
        iter26.Next("")
        iter27.Next("")
        db1.Set("ybglsmtq@34", "k")
        iter27.SeekPrefixGE("zqpknumxae@3")
        db1.IngestExternalFiles(external0, "tfja" /* start */, "ybglsmtq" /* end */, "@55" /* syntheticSuffix */, "" /* syntheticPrefix */, external0, "jaztfja" /* start */, "jazybglsmtq" /* end */, "" /* syntheticSuffix */, "jaz" /* syntheticPrefix */)
        db1.Get("prnyxfvr@8")
        iter27.Last()
        db1.Set("uxwoabffske@14", "h")
        iter26.First()
        iter26.SeekLT("mdwszgxqb@9", "")
        iter26.Prev("")
        iter23.SeekPrefixGE("aweambbwu")
        iter24.InternalNext()
        iter23.SeekGE("ybglsmtq@34", "")
        db1.IngestExternalFiles(external0, "sxowi" /* start */, "uhbbl" /* end */, "@57" /* syntheticSuffix */, "" /* syntheticPrefix */, external0, "stqapnmo" /* start */, "suxwoabffske" /* end */, "" /* syntheticSuffix */, "s" /* syntheticPrefix */, external0, "jshvit" /* start */, "kblx" /* end */, "@56" /* syntheticSuffix */, "" /* syntheticPrefix */, external0, "kblx" /* start */, "stqapnmo" /* end */, "" /* syntheticSuffix */, "" /* syntheticPrefix */)
        iter23.SeekPrefixGE("aweambbwu@9")
        iter25.Next("")
        iter23.SetBounds("mdwszgxqb@23", "ywgse@5")
        db1.Set("aweambbwu@12", "pplmnxxgfldgox")
        iter26.Close()
        iter27.Close()
        iter21.Close()
        iter23.Close()
        iter25.Close()
        iter24.Close()
        snap18.Close()
        snap19.Close()
        db1.Close()
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --compare "/artifacts/tmp/_tmp/366cdd6e86440d4e9e9c98d28c5c6592/TestMetaCrossVersion1541470195/001/1723281556929389535/master_791b3749_419129674420735526_008/_meta/240810-092056.1993592836014/{standard-000,random-014}" --try-to-reduce -v
        
        --- FAIL: TestMetaCrossVersion/execution-master/sha=7e594e28-seed=13102867164713446728-opts=random-012(sha=848f3c6e-seed=8168522583084423651-opts=standard-001(sha=1becf831-seed=2386158681653477521-opts=random-005(sha=ffd5ce9c-seed=16392966780622546391-opts=standard-003(<empty>)))) (281.58s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaCrossVersion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: PEBBLE-235

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -1415,6 +1415,13 @@ func (g *generator) writerIngestExternalFiles() {
 				continue
 			}
 
+			// We can only use a synthetic suffix if we don't have overlapping range
+			// key sets (because they will become logically conflicting when we
+			// replace their suffixes with the synthetic one).
+			if g.keyManager.ExternalObjectHasOverlappingRangeKeySets(objs[i].externalObjID) {
+				continue
+			}
+
 			// Generate a suffix that sorts before any previously generated suffix.
 			objs[i].syntheticSuffix = g.keyGenerator.IncMaxSuffix()
 		}

--- a/metamorphic/key_manager.go
+++ b/metamorphic/key_manager.go
@@ -218,6 +218,9 @@ type objKeyMeta struct {
 	hasRangeDels     bool
 	hasRangeKeys     bool
 	hasRangeKeyUnset bool
+	// List of RangeKeySets for this object. Used to check for overlapping
+	// RangeKeySets in external files.
+	rangeKeySets []pebble.KeyRange
 }
 
 // MergeKey adds the given key at the given meta timestamp, merging the histories as needed.
@@ -252,6 +255,7 @@ func (okm *objKeyMeta) MergeFrom(from *objKeyMeta, metaTimestamp int, cmp base.C
 	okm.hasRangeDels = okm.hasRangeDels || from.hasRangeDels
 	okm.hasRangeKeys = okm.hasRangeKeys || from.hasRangeKeys
 	okm.hasRangeKeyUnset = okm.hasRangeKeyUnset || from.hasRangeKeyUnset
+	okm.rangeKeySets = append(okm.rangeKeySets, from.rangeKeySets...)
 }
 
 // objKeyMeta looks up the objKeyMeta for a given object, creating it if necessary.
@@ -323,6 +327,25 @@ func (k *keyManager) KeysForExternalIngest(obj externalObjWithBounds) []keyMeta 
 		}
 	}
 	return res
+}
+
+func (k *keyManager) ExternalObjectHasOverlappingRangeKeySets(externalObjID objID) bool {
+	meta := k.objKeyMeta(externalObjID)
+	if len(meta.rangeKeySets) == 0 {
+		return false
+	}
+	ranges := meta.rangeKeySets
+	// Sort by start key.
+	slices.SortFunc(ranges, func(a, b pebble.KeyRange) int {
+		return k.comparer.Compare(a.Start, b.Start)
+	})
+	// Check overlap between adjacent ranges.
+	for i := 0; i < len(ranges)-1; i++ {
+		if ranges[i].OverlapsKeyRange(k.comparer.Compare, ranges[i+1]) {
+			return true
+		}
+	}
+	return false
 }
 
 func (k *keyManager) nextMetaTimestamp() int {
@@ -614,11 +637,17 @@ func (k *keyManager) update(o op) {
 		k.objKeyMeta(s.writerID).hasRangeKeys = true
 	case *rangeKeySetOp:
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
-		k.objKeyMeta(s.writerID).hasRangeKeys = true
+		meta := k.objKeyMeta(s.writerID)
+		meta.hasRangeKeys = true
+		meta.rangeKeySets = append(meta.rangeKeySets, pebble.KeyRange{
+			Start: s.start,
+			End:   s.end,
+		})
 	case *rangeKeyUnsetOp:
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
-		k.objKeyMeta(s.writerID).hasRangeKeys = true
-		k.objKeyMeta(s.writerID).hasRangeKeyUnset = true
+		meta := k.objKeyMeta(s.writerID)
+		meta.hasRangeKeys = true
+		meta.hasRangeKeyUnset = true
 	case *ingestOp:
 		// Some ingestion operations may attempt to ingest overlapping sstables
 		// which is prohibited. We know at generation time whether these

